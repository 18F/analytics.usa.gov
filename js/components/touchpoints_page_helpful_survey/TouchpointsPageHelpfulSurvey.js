import { useEffect, React } from "react";

/**
 * Creates the site TouchpointsPageHelpfulSurvey. Loads the JS code provided by
 * Touchpoints after the component is loaded. This Touchpoints code creates the
 * remainder of the survey form elements.
 *
 * @returns {import('react').ReactElement} The rendered element
 */
function TouchpointsPageHelpfulSurvey() {
  const initialFormDelayMilliseconds = 8000;

  useEffect(() => {
    document.addEventListener("onTouchpointsFormLoaded", (e) => {
      if (
        e.detail &&
        e.detail.formComponent
          .formComponent()
          .getAttribute("data-touchpoints-form-id") === "8fc3c209"
      ) {
        const formComponent = e.detail.formComponent;
        const formElement = formComponent.formComponent();
        _delayFormSubmit(formComponent, initialFormDelayMilliseconds);

        const touchpointsFooter = formElement.querySelector(
          "footer.touchpoints-footer-banner",
        );
        const yesButton = formElement.querySelector(
          'input.usa-radio__input[value="Yes"]',
        );
        const noButton = formElement.querySelector(
          'input.usa-radio__input[value="No"]',
        );
        const feedbackTextArea = formElement.querySelector("textarea");
        const questions = formElement.querySelectorAll(".questions .question");
        const header = formElement.querySelector("#fba-form-title-8fc3c209");
        const submitButton = formElement.querySelector(
          'button.submit_form_button[type="submit"]',
        );

        // Hide the touchpoints footer always
        if (touchpointsFooter) {
          touchpointsFooter.classList.add("hide");
        }

        // Hide the second question and submit button initially
        if (questions.length > 0 && submitButton) {
          questions[1].classList.add("hide");
          submitButton.classList.add("hide");
        }

        const handleYesNoButtonClick = () => {
          if (questions.length > 0 && header && submitButton) {
            // Hide the first form question and the header. Show the second along
            // with the submit button.
            questions[0].classList.add("hide");
            header.classList.add("hide");
            questions[1].classList.remove("hide");
            submitButton.classList.remove("hide");
          }
        };

        if (yesButton) {
          yesButton.addEventListener("click", handleYesNoButtonClick);
        }

        if (noButton) {
          noButton.addEventListener("click", handleYesNoButtonClick);
        }

        if (submitButton) {
          submitButton.addEventListener("click", () => {
            const eventDetails = {
              event_category: "cx_feedback",
              event_action: "was_this_page_helpful_v3.0",
            };

            if (feedbackTextArea.value) {
              eventDetails.reason = feedbackTextArea.value;
            }

            // Only send the event if the yes/no radio option has been set.
            if (yesButton.checked) {
              eventDetails.event_label = "yes";
              gas4("was_this_helpful_submit", eventDetails);
            } else if (noButton.checked) {
              eventDetails.event_label = "no";
              gas4("was_this_helpful_submit", eventDetails);
            }
          });
        }
      }
    });
    require("../../lib/touchpoints_page_helpful_survey");
  });

  return (
    <>
      <div id="touchpoints-yes-no-form"></div>
    </>
  );
}

function _delayFormSubmit(formComponent, delayMilliseconds) {
  /**
   * Set an initial timeout for the form. If the form is submitted too
   * quickly (because it's a bot), then sendFeedback is a no-op
   */
  const originalSendFeedbackFunction = formComponent.sendFeedback;
  /* eslint-disable no-empty */
  formComponent.sendFeedback = () => {};
  /* eslint-enable no-empty */

  setTimeout(() => {
    formComponent.sendFeedback = originalSendFeedbackFunction;
  }, delayMilliseconds);
}

TouchpointsPageHelpfulSurvey.propTypes = {};

export default TouchpointsPageHelpfulSurvey;
