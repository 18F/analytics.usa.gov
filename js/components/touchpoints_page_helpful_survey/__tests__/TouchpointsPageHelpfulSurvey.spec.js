import React from "react";
import { render } from "@testing-library/react";
import TouchpointsPageHelpfulSurvey from "../TouchpointsPageHelpfulSurvey";

jest.mock("export-to-csv", () => {
  return {};
});

describe("TouchpointsPageHelpfulSurvey", () => {
  let component;

  describe("initial render", () => {
    beforeEach(() => {
      component = render(<TouchpointsPageHelpfulSurvey />);
    });

    it("renders the form", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
