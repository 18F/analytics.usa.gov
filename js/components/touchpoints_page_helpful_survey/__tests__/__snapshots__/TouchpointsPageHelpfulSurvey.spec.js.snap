// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TouchpointsPageHelpfulSurvey initial render renders the form 1`] = `
<DocumentFragment>
  <div
    id="touchpoints-yes-no-form"
  >
      
    <div
      aria-label="Feedback modal dialog"
      aria-labelledby="fba-modal-title"
      aria-modal="false"
      class="fba-modal-dialog"
      role="dialog"
    >
      
    
      <div
        class="touchpoints-form-wrapper "
        data-touchpoints-form-id="8fc3c209"
        id="touchpoints-form-8fc3c209"
      >
        
  
        <div
          class="wrapper"
        >
          
    
          <div
            class="fba-alert usa-alert usa-alert--success"
            hidden=""
            role="status"
          >
            
  
            <div
              class="usa-alert__body"
            >
              
    
              <h2
                class="usa-alert__heading"
              >
                
      Success
    
              </h2>
              
    
              <div
                class="usa-alert__text"
              >
                
      Thank you. Your feedback has been received.
    
              </div>
              
  
            </div>
            

          </div>
          

          <div
            class="fba-alert-error usa-alert usa-alert--error"
            hidden=""
            role="alert"
          >
            
  
            <div
              class="usa-alert__body"
            >
              
    
              <h2
                class="usa-alert__heading"
              >
                
      Error
    
              </h2>
              
    
              <p
                class="usa-alert__text"
              >
                
      alert message
    
              </p>
              
  
            </div>
            

          </div>
          

    

          <form
            action="https://touchpoints.app.cloud.gov/touchpoints/8fc3c209/submissions.json"
            class="usa-form usa-form--large margin-bottom-2"
            method="POST"
          >
            
  
            <div
              class="touchpoints-form-body"
            >
              
      
              <div
                class="section fba-visible"
              >
                



        
                <div
                  class="questions"
                >
                  

          
                  <div
                    class="question white-bg"
                    id="question_44546"
                  >
                    
              
                    <fieldset
                      class="usa-fieldset radios margin-top-3"
                    >
                      
  
                      <legend
                        class="usa-legend"
                        id="question-label-44546"
                      >
                        
 
                        <h1
                          class="word-break fba-modal-title"
                          id="fba-form-title-8fc3c209"
                        >
                          
    Is this page helpful?
  
                        </h1>
                        

                      </legend>
                      

  
                      <div
                        class="question-options"
                      >
                        
    
                        <div
                          class="radio-button usa-radio question-option"
                          data-id="66480"
                        >
                          
      
                          <input
                            class="usa-radio__input usa-radio__input--tile"
                            id="question_option_66480"
                            name="question_44546_answer_01"
                            type="radio"
                            value="Yes"
                          />
                          
      
                          <label
                            class="usa-radio__label"
                            for="question_option_66480"
                          >
                            Yes
                          </label>
                          
    
                        </div>
                        
    
                        <div
                          class="radio-button usa-radio question-option"
                          data-id="66481"
                        >
                          
      
                          <input
                            class="usa-radio__input usa-radio__input--tile"
                            id="question_option_66481"
                            name="question_44546_answer_01"
                            type="radio"
                            value="No"
                          />
                          
      
                          <label
                            class="usa-radio__label"
                            for="question_option_66481"
                          >
                            No
                          </label>
                          
    
                        </div>
                        
  
                      </div>
                      

                    </fieldset>
                    

          
                  </div>
                  

          
                  <div
                    class="question white-bg hide"
                    id="question_50470"
                  >
                    
              
                    <div
                      role="group"
                    >
                      
  
                      <label
                        class="usa-label"
                        for="question_50470_answer_02"
                      >
                        
  How can we make this page better?

                      </label>
                      

  
                      <textarea
                        class="usa-textarea"
                        id="question_50470_answer_02"
                        maxlength="10000"
                        name="question_50470_answer_02"
                      />
                      

                    </div>
                    

          
                  </div>
                  
        
                </div>
                

          
                <button
                  class="usa-button submit_form_button hide"
                  type="submit"
                >
                  Submit
                </button>
                
      
              </div>
              
        
              <input
                autocomplete="off"
                id="fba_location_code"
                name="fba_location_code"
                tabindex="-1"
                type="hidden"
              />
              
    
              <input
                aria-hidden="true"
                autocomplete="off"
                class="display-none"
                id="fba_directive"
                name="fba_directive"
                tabindex="-1"
                title="fba_directive"
                type="text"
              />
              
  
            </div>
            

          </form>
          

  
        </div>
        
  
      

        <section
          aria-label="footer An official website of the United States government"
          class="usa-banner"
        >
          
  
          <footer
            class="usa-banner__header touchpoints-footer-banner hide"
          >
            
    
            <div
              class="usa-banner__inner"
            >
              
      
              <div
                class="grid-col-auto"
              >
                
        
                <img
                  alt="U.S. flag"
                  aria-hidden="true"
                  class="usa-banner__header-flag"
                  src="https://touchpoints.app.cloud.gov/img/us_flag_small.png"
                />
                
      
              </div>
              
      
              <div
                class="grid-col-fill tablet:grid-col-auto"
              >
                
        
                <p
                  class="usa-banner__header-text"
                >
                  
          An official form of the United States government.
          Provided by
          
                  <a
                    class="usa-link--external"
                    href="https://touchpoints.digital.gov"
                    rel="noopener"
                    target="_blank"
                  >
                    Touchpoints
                  </a>
                  
          
                  <br />
                  

        
                </p>
                
      
              </div>
              
    
            </div>
            
  
          </footer>
          

        </section>
        



      </div>
      
  
    </div>
    

  </div>
</DocumentFragment>
`;
